src/driver/esp32_rmt.rs

```diff
@@ -1,150 +1,230 @@
     ///
     /// # Errors
     ///
     /// Returns an error if the RMT driver initialization failed.
-    pub fn new(channel_num: u8, gpio_num: u32) -> Result<Self, Ws2812Esp32RmtDriverError> {
-        let channel = channel_num as rmt_channel_t;
-        let gpio_num = gpio_num as gpio_num_t;
-        let clk_div = 2;
-
-        let rmt_cfg = rmt_config_t {
-            rmt_mode: rmt_mode_t_RMT_MODE_TX,
-            channel,
-            gpio_num,
-            clk_div,
-            mem_block_num: 1,
-            __bindgen_anon_1: rmt_config_t__bindgen_ty_1 {
-                tx_config: rmt_tx_config_t {
-                    loop_en: false,
-                    carrier_level: rmt_carrier_level_t_RMT_CARRIER_LEVEL_HIGH,
-                    carrier_en: false,
-                    idle_level: rmt_idle_level_t_RMT_IDLE_LEVEL_LOW,
-                    idle_output_en: true,
-                    ..Default::default()
-                },
-            },
-            ..Default::default()
-        };
-        esp!(unsafe { rmt_config(&rmt_cfg) })?;
-        esp!(unsafe { rmt_driver_install(channel, 0, 0) })?;
-        esp!(unsafe { rmt_translator_init(channel, Some(ws2812_rmt_adapter)) })?;
+    pub fn new<C: RmtChannel>(
+        channel: impl Peripheral<P = C> + 'd,
+        pin: impl Peripheral<P = impl OutputPin> + 'd,
+    ) -> Result<Self, Ws2812Esp32RmtDriverError> {
+        let config = TransmitConfig::new().clock_divider(1);
+        let tx = TxRmtDriver::new(channel, pin, &config)?;
 
-        let _encoder =
-            WS2812_ITEM_ENCODER.get_or_try_init(|| Ws2812Esp32RmtItemEncoder::new(channel))?;
+        let clock_hz = tx.counter_clock()?;
+        let encoder = Ws2812Esp32RmtItemEncoder::new(clock_hz)?;
 
-        Ok(Self { channel })
+        Ok(Self { tx, encoder })
     }
 
-    /// Writes pixel data from the slice to the IO pin.
+    /// Writes pixel data from a pixel-byte-pointer sequence to the IO pin.
     ///
     /// Byte count per LED pixel and channel order is not handled by this method.
-    /// The data has to be correctly laid out in the slice depending on the LED strip model.
+    /// The pixel data sequence has to be correctly laid out depending on the LED strip model.
     ///
     /// # Errors
     ///
     /// Returns an error if an RMT driver error occurred.
     ///
-    /// # Panics
+    /// # Warning
     ///
-    /// Panics if the given slice is longer than `u32::MAX`.
-    pub fn write(&mut self, pixel_data: &[u8]) -> Result<(), Ws2812Esp32RmtDriverError> {
-        let data_ptr = pixel_data.as_ptr();
-        esp!(unsafe { rmt_write_sample(self.channel, data_ptr, pixel_data.len(), true) })?;
+    /// Iteration of `pixel_sequence` happens inside an interrupt handler so beware of side-effects
+    /// that don't work in interrupt handlers.
+    /// See [esp_idf_hal::rmt::TxRmtDriver#start_iter_blocking()] for details.
+    pub fn write_ptr_iter_blocking<'a, 'b, T>(
```

Comment:
What part of this method deals with "pointers"? I don't understand the documentation or method name really. You mean the fact that the iterator returns `&u8`? That would more more idiomatically be called `ref` in Rust.

Why is this method needed? It just "clones" the bytes anyway and then behaves the same way as `write_val_iter_blocking`. Does this deliver considerable ergonomics I fail to see? I would probably opt for having just one method and have the caller append `.copied()` to their iterator if needed.

Reply:
Thanks for the review!

Surely just `Iterator<u8>` should be handled as you suggest. Then, the name becomes simple, since we only need to care about whether the name is blocking or not.